File: archived/TextSegmentation/js/js/scenario1-ExtractTextSegments.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";
    var page = WinJS.UI.Pages.define("/html/scenario1-ExtractTextSegments.html", {
        ready: function (element, options) {
            document.getElementById("WordSegmentButton").addEventListener("click", showWordSegments, false);
            document.getElementById("SelectionSegmentButton").addEventListener("click", showSelectionSegments, false);
        }
    });

    function showWordSegments() {

        // Initialize and obtain input values
        var notifyText = "";

        // Obtain the input string value, check for non-emptiness
        var inputStringBox = document.getElementById("inputStringBox");
        var inputStringText = inputStringBox.value;
        if ("" === inputStringText) {
            notifyText = "Cannot compute word segments: input string is empty.";
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }

        // Obtain the language tag value, check for non-emptiness and that it is valid
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        var languageTagBox = document.getElementById("languageTagBox");
        var languageTagText = languageTagBox.value;
        if ("" === languageTagText) {
            notifyText += "Language tag input is empty ... using generic-language segmentation rules.";
            languageTagText = "und";
        }
        else {
            if (!Windows.Globalization.Language.isWellFormed(languageTagText)) {
                notifyText = "Language tag is not well formed: \"" + languageTagText + "\"";
                WinJS.log && WinJS.log(notifyText, "sample", "error");
                return;
            }
        }

        // Notify that we are going to calculate word segments
        notifyText += "\n\nCalculating word segments ...\n";
        notifyText += "\nInput: \"" + inputStringText + "\"";
        notifyText += "\nLanguage Tag: \"" + languageTagText + "\"";

        // Construct the WordsSegmenter instance
        var segmenter = new Windows.Data.Text.WordsSegmenter(languageTagText);

        // Tokenize the word segments, iterate over them and add to notifyText
        var wordSegments = segmenter.getTokens(inputStringText);
        wordSegments.forEach(
            function (wordSegment) {
                notifyText += "\n\tWord Bound---->\"" + wordSegment.text + "\"";
                var alternateForms = wordSegment.alternateForms;
                alternateForms.forEach(
                    function (alternateForm) {
                        notifyText += "\n\t\tAlternate---->\"" + alternateForm.alternateText + "\"";
                    }
                );
            }
        );

        // send notifyText to the output pane
        WinJS.log && WinJS.log(notifyText, "sample", "status");
    }

    function showSelectionSegments() {

        // Initialize and obtain input values
        var notifyText = "";

        // Obtain the input string value, check for non-emptiness
        var inputStringBox = document.getElementById("inputStringBox");
        var inputStringText = inputStringBox.value;
        if ("" === inputStringText) {
            notifyText = "\nCannot compute selection segments: input string is empty.";
            WinJS.log && WinJS.log(notifyText, "sample", "error");
            return;
        }

        // Obtain the language tag value, check for non-emptiness and that it is valid
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        var languageTagBox = document.getElementById("languageTagBox");
        var languageTagText = languageTagBox.value;
        if ("" === languageTagText) {
            notifyText += "\nLanguage tag input is empty ... using generic-language segmentation rules.";
            languageTagText = "und";
        }
        else {
            if (!Windows.Globalization.Language.isWellFormed(languageTagText)) {
                notifyText = "\nLanguage tag is not well formed: \"" + languageTagText + "\"";
                WinJS.log && WinJS.log(notifyText, "sample", "error");
                return;
            }
        }

        // Notify that we are going to calculate selection segments
        notifyText += "\n\nCalculating selection segments ...\n";
        notifyText += "\nInput: \"" + inputStringText + "\"";
        notifyText += "\nLanguage Tag: \"" + languageTagText + "\"";

        // Construct the SelectableWordsSegmenter instance
        var segmenter = new Windows.Data.Text.SelectableWordsSegmenter(languageTagText);

        // Tokenize the selection segments, iterate over them and add to notifyText
        var selectionSegments = segmenter.getTokens(inputStringText);
        selectionSegments.forEach(
            function (selectionSegment) {
                notifyText += "\n\tSelection Bound---->\"" + selectionSegment.text + "\"";
            }
        );

        // send notifyText to the output pane
        WinJS.log && WinJS.log(notifyText, "sample", "status");
    }
})();

File: Samples/TextSegmentation/cppwinrt/Scenario1_ExtractTextSegments.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario1_ExtractTextSegments.h"
#include "Scenario1_ExtractTextSegments.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;

namespace winrt::SDKTemplate::implementation
{
    Scenario1_ExtractTextSegments::Scenario1_ExtractTextSegments()
    {
        InitializeComponent();
    }

    // This is the click handler for the 'Word Segments' button.
    //
    // When this button is activated, the Text Segmentation API will calculate
    // the word segments from the given input string and language tag, and then
    // print out the word and alternate-form segments for that language.
    void Scenario1_ExtractTextSegments::WordSegmentsButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        std::wostringstream result;

        // Obtain the input string value, check for non-emptiness
        hstring inputStringText = inputStringBox().Text();
        if (inputStringText.empty())
        {
            rootPage.NotifyUser(L"Cannot compute word selection segments: input string is empty.", NotifyType::ErrorMessage);
            return;
        }

        // Obtain the language tag value, check for non-emptiness
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        hstring languageTagText = languageTagBox().Text();
        if (languageTagText.empty())
        {
            result << L"Language tag input is empty ... using generic-language segmentation rules.\n";
            languageTagText = L"und";   // This is used for non language-specific locales. 'und' is short for 'undetermined'.
        }
        else
        {
            if (!Windows::Globalization::Language::IsWellFormed(languageTagText))
            {
                rootPage.NotifyUser(L"Language tag is not well formed: \"" + languageTagText + L"\"", NotifyType::ErrorMessage);
                return;
            }
        }

        // Notify that we are going to calculate word segments
        result
            << L"\nCalculating word segments ...\n\n"
            << L"Input: \"" << std::wstring_view(inputStringText) << L"\"\n"
            << L"Language Tag: \"" << std::wstring_view(languageTagText) << L"\"\n";

        // Construct the WordsSegmenter instance
        Windows::Data::Text::WordsSegmenter segmenter(languageTagText);
        auto wordSegments = segmenter.GetTokens(inputStringText);

        // Iterate over the tokenized segments
        for (auto&& wordSegment : wordSegments)
        {
            result << L"\tWord---->\"" << std::wstring_view(wordSegment.Text()) << L"\"\n";
            for  (auto&& alternateForm : wordSegment.AlternateForms())
            {
                result << L"\t\tAlternate---->\"" << std::wstring_view(alternateForm.AlternateText()) << L"\"\n";
            }
        }

        // Set results to output box
        rootPage.NotifyUser(result.str(), NotifyType::StatusMessage);
    }

    // This is the click handler for the 'Selection Segments' button.
    //
    // When this button is activated, the Text Segmentation API will calculate
    // the selection segments from the given input string and language tag, and then
    // print out the word and alternate-form segments for that language.
    //
    // Selection segments differ from word segments in that they describe the bounds
    // between active-selection snapping behavior.
    void Scenario1_ExtractTextSegments::SelectionSegmentsButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        std::wostringstream result;

        // Obtain the input string value, check for non-emptiness
        hstring inputStringText = inputStringBox().Text();
        if (inputStringText.empty())
        {
            rootPage.NotifyUser(L"Cannot compute selection segments: input string is empty.", NotifyType::ErrorMessage);
            return;
        }

        // Obtain the language tag value, check for non-emptiness
        // Ex. Valid Values:
        //     "en-US" (English (United States))
        //     "fr-FR" (French (France))
        //     "de-DE" (German (Germany))
        //     "ja-JP" (Japanese (Japan))
        //     "ar-SA" (Arabic (Saudi Arabia))
        //     "zh-CN" (China (PRC))
        hstring languageTagText = languageTagBox().Text();
        if (languageTagText.empty())
        {
            result << L"Language tag input is empty ... using generic-language segmentation rules.\n";
            languageTagText = L"und";   // This is used for non language-specific locales. 'und' is short for 'undetermined'.
        }
        else
        {
            if (!Windows::Globalization::Language::IsWellFormed(languageTagText))
            {
                rootPage.NotifyUser(L"Language tag is not well formed: \"" + languageTagText + L"\"", NotifyType::ErrorMessage);
                return;
            }
        }

        // Notify that we are going to calculate selection segments
        result
            << L"\nCalculating selection segments ...\n\n"
            << L"Input: \"" << std::wstring_view(inputStringText) << L"\"\n"
            << L"Language Tag: \"" << std::wstring_view(languageTagText) << L"\"\n";

        // Construct the SelectableWordsSegmenter instance
        Windows::Data::Text::SelectableWordsSegmenter segmenter(languageTagText);
        auto selectionSegments = segmenter.GetTokens(inputStringText);

        // Iterate over the tokenized segments
        for (auto&& selectionSegment : selectionSegments)
        {
            result << L"\tWord---->\"" << std::wstring_view(selectionSegment.Text()) << L"\"\n";
        }

        // Set results to output box
        rootPage.NotifyUser(result.str(), NotifyType::StatusMessage);
    }
}
